pytorch_lightning/loops/base.py

```diff
@@ -24,8 +24,10 @@
 from pytorch_lightning.utilities.apply_func import apply_to_collection
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
+T = TypeVar("T")  # the output type of `run`
 
-class Loop(ABC):
+
+class Loop(ABC, Generic[T]):
```

Comment:
what if we hypothetically also wanted to express that the output of on advance is the input to on_advance_end but not the same as for on_run_end/run? Would you inherit here from another generic?

Reply:
This seems to work:

```python
from typing import TypeVar, Generic

A = TypeVar("A")
B = TypeVar("B")


class Base(Generic[A, B]):
    def run(self) -> A:
        return self.on_run_end()

    def advance(self) -> B:
        ...

    def on_advance_end(self, b: B) -> None:
        ...

    def on_run_end(self) -> A:
        ...


class Foo:
    ...


class Bar:
    ...


class Child(Base[Foo, Bar]):
    def end(self) -> Foo:
        ...

    def on_advance_end(self, b: Bar) -> None:
        ...
```